modules/ListView.js

```diff
@@ -431,27 +512,71 @@
      */
-    __createChildView: function(model) {
-      var childView,
-        ChildViewClass = this.childView;
-      if (!_.isFunction(this.childView.extend)) {
-        ChildViewClass = this.childView(model);
+    __updateByAddingRemoving: function(oldViews, newViews, staleViews) {
+      var firstChildViewLeft, injectionSite,
+        view = this,
+        sizeOfOldViews = _.size(oldViews),
+        sizeOfNewViews = _.size(newViews),
+        sizeOfStaleViews = _.size(staleViews);
+      if (view.childrenContainer && sizeOfOldViews && sizeOfOldViews == sizeOfStaleViews) {
+        // we removed all the views!
+        injectionSite = $('<span>');
+        _.first(oldViews).$el.before(injectionSite);
       }
-      childView = new ChildViewClass(this.__generateChildArgs(model));
-      this.registerTrackedView(childView, { shared: false });
-      this.__modelToViewMap[model[this.__modelId]] = childView.cid;
-      this.trigger('child-view-added', {model: model, view: childView});
-      return childView;
+      _.each(staleViews, function(staleViewInfo, indexOfView) {
+        _removeItemView.call(view, staleViewInfo.view, staleViewInfo.modelId);
+      });
+      _.each(newViews, function(createdViewInfo, indexOfView) {
+        if (createdViewInfo.indexOfModel === 0) {
+          // need to handle this case uniquely.
+          var replaceMethod;
+          if (!view.childrenContainer) {
+            replaceMethod = _.bind(view.$el.prepend, view.$el);
+          } else {
+            if (injectionSite) {
+              replaceMethod = _.bind(injectionSite.replaceWith, injectionSite);
+            } else {
+              var staleModelIdMap = _.indexBy(staleViews, 'modelId');
+              var firstModelIdLeft = _.find(view.__orderedModelIdList, function(modelId) {
+                return !staleModelIdMap[modelId];
+              });
+              firstChildViewLeft = view.getChildView(view.__modelToViewMap[firstModelIdLeft]);
+              replaceMethod = _.bind(firstChildViewLeft.$el.prepend, firstChildViewLeft.$el);
+            }
+          }
+          view.attachView(null, createdViewInfo.view, {
+            replaceMethod: replaceMethod,
+            discardInjectionSite: true
+          });
+        } else {
+          // There will always the view before this one because we are adding new views in order
+          // and we took care of the initial case.
+          _addItemView.call(view, createdViewInfo.view, createdViewInfo.indexOfModel);
+        }
+      });
+      this.reorder();
+    },
+
+    /**
+     * Updates the internal list of model ids that correspond to the models used for the current
+     * list of item views. The order is the same order of the item views.
+     * @method __updateOrderedModelIdList
+     * @private
+     */
+    __updateOrderedModelIdList: function() {
+      this.__orderedModelIdList = _.pluck(this.modelsToRender(), this.__modelId);
```

Comment:
this is really a snapshot of the model order at this time?  Maybe that would be a clearer naming?  orderedModelIdList looks like it should be a live reflection of modelsToRender()'s order.  Which it isn't (it just only valid once this method is called if modelsToRender() changes).